maths/bbp_formula.py

```diff
@@ -0,0 +1,47 @@
+def bbp_pi_approximation(num_iterations: int = 1000) -> float:
+    """
+    Uses a popular pi-approximating algorithm known as the 
+    Bailey-Borwein-Plouffe (BBP) formula to approximate pi.
+    Wikipedia page:
+    https://en.wikipedia.org/wiki/Bailey%E2%80%93Borwein%E2%80%93Plouffe_formula
+    @param num_iterations: nonnegative integer representing the number 
+    of terms in the formula to sum (more iterations leads to more accuracy)
+    @return: the approximation of pi given by the formula
+    
+    >>> bbp_pi_approximation(-10)
+    Traceback (most recent call last):
+      ...
+    ValueError: Please input a nonnegative integer
+    >>> bbp_pi_approximation(1.7)
+    Traceback (most recent call last):
+      ...
+    ValueError: Please input a nonnegative integer
+    >>> "%.8f" % bbp_pi_approximation()
+    '3.14159265'
+    >>> "%.8f" % bbp_pi_approximation(1000)
+    '3.14159265'
+    >>> "%.8f" % bbp_pi_approximation(0)
+    '3.13333333'
+    >>> "%.8f" % bbp_pi_approximation(1)
+    '3.14142247'
+    """
+    if (not isinstance(num_iterations, int)) or num_iterations < 0:
+        raise ValueError("Please input a nonnegative integer")
+    pi_approx = 0.0
+    exponential_part = 1.0
+    # num_iterations + 1 because sum notation includes the last term
+    for index in range(num_iterations + 1):
```

Comment:
I'm not sure the + 1 part actually is needed. By definition this is an infinite sum which just gets more and more accurate with each iteration, so anytime you stop is technically the last iteration :) 
And it might fit better to how you define `num_iterations` as well. As an example, default value for `num_iterations` is 1000, and starting with 0 and ending with 999 gives you a total of 1000 iterations.

Very minor detail though and shouldn't affect the final result.

Reply:
I tried to design it so that num_iterations represented the number on top of the sigma expression in the formula. Now that I think about it, though, it might have been better to name the variable something like upper_limit_of_summation rather than num_iterations